pandas 함수를 이어서 공부
정렬, Aggregation 함수, Groupby 적용
DataFrame, Series의 정렬을 위해서 sort_values()메서드 이용한다.


주요 파라미터는 by, ascending, inplace가 있다.
1. by: 특정 칼럼을 입력하면 해당 칼럼으로 정렬을 수행한다.
2. ascending: ascending=True가 기본값이며 오름차순으로 정렬된다.
3. inplace: inplace=False가 기본값이며 호출된 Dataframe은 그대로이며 정렬된 DataFrame을 결과로 반환한다.

In [28]:
import numpy as np
import pandas as pd

In [29]:
titanic=pd.read_csv('train.csv')
titanic.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [30]:
#정렬
itanic_sorted=titanic.sort_values(by=['Name'])
titanic_sorted.head(3)

PassengerId  Survived  Pclass                              Name     Sex  \
845          846         0       3               Abbing, Mr. Anthony    male   
746          747         0       3       Abbott, Mr. Rossmore Edward    male   
279          280         1       3  Abbott, Mrs. Stanton (Rosa Hunt)  female   

      Age  SibSp  Parch     Ticket   Fare Cabin Embarked  
845  42.0      0      0  C.A. 5547   7.55   NaN        S  
746  16.0      1      1  C.A. 2673  20.25   NaN        S  
279  35.0      1      1  C.A. 2673  20.25   NaN        S

In [31]:
#Aggregation: min(),max(),sum(), count()와 같은 Aggregation 함수는 모든 칼럼에 aggregation을 적용한다.
titanic.count()

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

In [32]:
#특정 부분만 추출해서 적용할 수도 있다.
titanic[['Age']].mean()

Age    29.699118
dtype: float64

In [33]:
#(by='Pclass')를 호출하면 Pclass 를 기준으로 groupby된 dataframe group 객체를 반환한다.
titanic_group=titanic.groupby(by='Pclass')
titanic_group.head(3)


PassengerId  Survived  Pclass  \
0             1         0       3   
1             2         1       1   
2             3         1       3   
3             4         1       1   
4             5         0       3   
6             7         0       1   
9            10         1       2   
15           16         1       2   
17           18         1       2   

                                                 Name     Sex   Age  SibSp  \
0                             Braund, Mr. Owen Harris    male  22.0      1   
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                              Heikkinen, Miss. Laina  female  26.0      0   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                            Allen, Mr. William Henry    male  35.0      0   
6                             McCarthy, Mr. Timothy J    male  54.0      0   
9                 Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
15                   Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
17                       Williams, Mr. Charles Eugene    male   NaN      0   

    Parch            Ticket     Fare Cabin Embarked  
0       0         A/5 21171   7.2500   NaN        S  
1       0          PC 17599  71.2833   C85        C  
2       0  STON/O2. 3101282   7.9250   NaN        S  
3       0            113803  53.1000  C123        S  
4       0            373450   8.0500   NaN        S  
6       0             17463  51.8625   E46        S  
9       0            237736  30.0708   NaN        C  
15      0            248706  16.0000   NaN        S  
17      0            244373  13.0000   NaN        S

In [34]:
#groupby: DataFrame에 있는 대상 칼럼을 제외하고 모든 칼럼이 해당 aggregation 함수를 적용한다.
titanic_group=titanic.groupby('Pclass').count()
titanic_group

PassengerId  Survived  Name  Sex  Age  SibSp  Parch  Ticket  Fare  \
Pclass                                                                      
1               216       216   216  216  186    216    216     216   216   
2               184       184   184  184  173    184    184     184   184   
3               491       491   491  491  355    491    491     491   491   

        Cabin  Embarked  
Pclass                   
1         176       214  
2          16       184  
3          12       491

In [35]:
titanic_group=titanic.groupby('Pclass')[['PassengerId','Survived']].count()
titanic_group

PassengerId  Survived
Pclass                       
1               216       216
2               184       184
3               491       491

In [40]:
#여러개의 함수명을 dataframegroupby 객체의 agg() 내에 인자로 입력해서 사용한다.
titanic.groupby('Pclass')['Age'].agg([max,min])

max   min
Pclass            
1       80.0  0.92
2       70.0  0.67
3       74.0  0.42

In [39]:
#여러개의 칼럼이 서로 다른 aggregation 함수흫 groupby에 호출하려면 agg()내에 입력 값이 딕셔너리 형식이여야 한다.
agg_format={'Age':'max', 'SibSp':'sum', 'Fare':'mean'}
titanic.groupby('Pclass').agg(agg_format)

Age  SibSp       Fare
Pclass                        
1       80.0     90  84.154687
2       70.0     74  20.662183
3       74.0    302  13.675550

결손 데이터 처리하기

결손 데이터: 칼럼에 값이 없는 즉, null값인 경우를 의미하며 넘파이인 경우에 NAN으로 표시한다.
1. isnan()으로 결손 데이터 여부 확인
: 모든 칼럼의 값을 NAN인지 False나 True로 알려준다.
2. fillna()로 결손 데이터 대체하기
: inplace=True를 추가하면 실제 데이터 세트값이 달라진다.

In [41]:
titanic.isna().head(3)

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket  \
0        False     False   False  False  False  False  False  False   False   
1        False     False   False  False  False  False  False  False   False   
2        False     False   False  False  False  False  False  False   False   

    Fare  Cabin  Embarked  
0  False   True     False  
1  False  False     False  
2  False   True     False

In [42]:
#sum()을 추가하면 True가1 로 Ralse가 2로 표시가 되어서 결손 데이터의 개수를 구할수 있다.
titanic.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [43]:
titanic['Cabin']=titanic['Cabin'].fillna('C000')
titanic.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500  C000        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250  C000        S

In [45]:
titanic['Age']=titanic['Age'].fillna(titanic['Age'].mean())
titanic['Embarked']=titanic['Embarked'].fillna('S')
titanic.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

In [46]:
#apply lambda 식으로 데이터 가공 : 함수를 만드는 대신에 사용가능
lambda_square=lambda x : x**2
print('3의 제곱은:', lambda_square(3))

3의 제곱은: 9


In [48]:
#여러개의 값을 입력 인자로 사용해야 할때는 map() 함수를 결해서 사용한다.
a=[1,2,3]
squares=map(lambda x: x**2, a)
list(squares)

[1, 4, 9]

In [60]:
#lambda 를 이용해서  Name_len을 생성해보자
titanic['Name_len']=titanic['Name'].apply(lambda x : len(x))
titanic[['Name','Name_len']].head(3)

Name  Name_len
0                            Braund, Mr. Owen Harris        23
1  Cumings, Mrs. John Bradley (Florence Briggs Th...        51
2                             Heikkinen, Miss. Laina        22

In [61]:
#lambda 식에 if else 절을 사용한다.
titanic['Child_Adult']=titanic['Age'].apply(lambda x : 'Child'if  x <=15 else 'Adult')
titanic[['Age','Child_Adult']].head(8)
#else if는 따로 지정되지 않아서 else안에 if else절을 중복해서 사용해야 한다.

Age Child_Adult
0  22.000000       Adult
1  38.000000       Adult
2  26.000000       Adult
3  35.000000       Adult
4  35.000000       Adult
5  29.699118       Adult
6  54.000000       Adult
7   2.000000       Child

사이킷런:
1.  파이썬 기반의 다른 머신러닝 페키지도 사이킷런 스타일의 API를 지향할 정도로 쉽고 가장 파이썬 스러운 API를 제공한다.
2. 머신러닝을 위헤 매우 다양한 알고리즘과 개발을 위한 편리한 프레임워크와 API를 제공한다.
3. 오랜 기간 실전 환경에서 검증되었은며 매우 많은 환경에서 사용되는 성숙한 라이브러리이다.

붓꽃 품종 예측하기

분류: 대표적인 지도학습 방법의 하나이다.
    지도학습은 명확한 정답이 주어진 데이터를 먼저 학습한 뒤 미지의 정답을 예측하는 방식이다.
    학습을 위해 주어진 데이터 세트를 테스트 데이터 세트로 지칭한다.
    
사이킷런 페키지내의 모듈명은 sklearn으로 시작하는 명명규칙이 있다.
 1. sklearn.datasets 내의 모듈은 사이킷런에서 자체적으로 제공하는 데이터 세트를 생성하는 모듈의 모임이다.
 2. sklearn.tree 내의 모듈은 트리 기반 ML알고리즘을 구현한 클래스의 모임이다.
 3. sklearn.model_selection 은 학습 데이터와 검증 데이터, 
    예측 데이터로 데이터를 분리하거나 최적의 하이퍼 파라미터로 평가하기 위한 다양한 모듈의 모임이다.
 4. 하이퍼 파라미터: 머신러닝 알고리즘 별로 최적의 학습을 위해 직접 입력하는 데이터를 통칭하며, 하이퍼 파라미터를 통헤 머신러닝 알고리즘의 성능을 튜닝할수 있디.

In [66]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split

In [70]:
#load_iris()함수를 이용하여 붓꽃 데이터 세트를 로딩한 후,
#피처들과 데이터 값이 어떻게 구성되어 있는지 확인하기 위해 DataFrame으로 변환한다.
import pandas as pd

#붓꽃 데이터세트를 로딩한다.
iris=load_iris()

#iris.data는 iris 데이터 세트에서 피처만으로 된 데이터를 numpy로 가지고 있다.
iris_data=iris.data

#iris,target 은 붓꽃 데이터 세트에서 레이블(결정 값) 데이터를 numpy로 가지고 있다.
iris_label=iris.target
print('iris target 값:', iris_label)
print('iris target 명:', iris.target_names)

#붓꽃 데이터 세트를 자세히 보기 위해 DataFrame으로 변환한다.
iris_df=pd.DataFrame(data=iris_data, columns=iris.feature_names)
iris_df['label']=iris.target
iris_df.head(3)

iris target 값: [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1
 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 2
 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2 2
 2 2]
iris target 명: ['setosa' 'versicolor' 'virginica']


sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   

   label  
0      0  
1      0  
2      0

결정 값(lable)은 0,1,2의 값으로 되어 있으며, setosa=1 versicolor=2, viginica=3 이다.

학습용 데이터와 테스트용 데이터는 반드시 분리가 되어야한다.

학습 데이터로 학습된 데이터가 얼마나 뛰어나 성능을 가지는지 평가하려면 데이터 세트가 필요하기 때문이다.

때문에 사이킷런은 train_test_split() API를 제공한다.
        
        이것을 이용하면 학습 데이터와 테스트 데이터를 test_size 파라미터 입력 값의 비율로 쉽게 분할한다.
        예를 들면 test_size=0.2로 입력 파라미터를 설정하면 전체 데이터 중 테스트 데이터가 20%,
        학습 데이터가 80%로 제이터를 분할한다.
        파라미터는 매개 변수로 두개 이상의 변수 사이의 함수를 간접적으로 표시할때 사용하는 함수이다.

In [71]:
x_train, x_test, y_train, y_test=train_test_split(iris_data, iris_label, test_size=0.2, random_state=11)

1. train_test_split()의 첫번째 파라미터인 iris_data는 피처 데이터 세트이다.
2. 두번째 파라미터인 iris_label은 레이블 데이터 세트이다.
3. test_size=0.2는 전체 데이터 세트 중 테스트 세트의 비율이다.
4. random_state는 호출할 때마다 같은 학습/테스트 용 데이터를 생성하기 위새 주어지는 난수 발생 값이다.
            
            train_test_split()는 호출시에 무작위로 데이터를 분리하므로 random_state를 지정하지 않으면 다른 데이터를 만든다.
            때문에 수행할 때마다 동일한 데이터 세트를 분리하기 위해 random_state를 일정한 숫자값으로 부여한다.

In [75]:
#사이킷런의 의사결정 트리 클래스인 DecisionTreeClassifier를 객체로 생성한다.
#fit()메서드에 학습용 피처 데이터 속성과 결정값 데이터 세트를 입력해 호출하면 학습을 수행한다.
df_clf=DecisionTreeClassifier(random_state=11)
df_clf.fit(x_train, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=11, splitter='best')

In [76]:
#predic()메서드에 테스트용 피처 데이터를 입력해 호출하면 학습된 모델 기반에서 테스트 데이터 세트에 대한 예측값을 반환하게 된다.
pred=df_clf.predict(x_test)

In [80]:
#정확도: 예측결과가 실제 레이블 값과 얼마나 정확하게 맞는지 평가하는 지표이다.
#사이킷런은 정확도 측정을 위해 accuracy_score()함수를 제공한다.
#첫번째 파라미터는 실제 레이블 데이터 세트, 두번빼 레이블은 예측 레이블 데이터 세트를 입력하면 된다.
from sklearn.metrics import accuracy_score
print('예측 정확도: {0:.4f}'.format(accuracy_score(y_test,pred)))

예측 정확도: 0.9333


예측 정확도가 93.33%로 추정이 된다.

붓꽃 데이터 세트로 분류를 예측한 프로세스를 정리하면 다음과 같다.
1. 데이터 세트 분리: 데이터를 학습 데이터와 테스트 데이터로 분류한다.
2. 모델 학습: 학습 데이터를 기반으로 ML 알고리즘을 적용해 모델을 학습시킨다.
3. 예측 수행: 학습된 ML모델을 이용애 테스트 데이터의 분류를 예측한다.
4. 평가: 이렇게 예측된 결과값과 테스트 데이터의 실제 결과값을 비교해 ML모델 성능을 평가한다.

사이킷런의 기반 프레임워크 익히기
1. 사이킷런은 ML모델 학습을 위해서 fit()을, 학습된 모델의 예측을 위해 predict()메서드를 제공한다.
2. 사이킷런에서 분류 알고리즘을 구현한 클래스를 Classifier로 , 회귀 알고리즘을 구현한 클래스를 Refressor로 지칭한다.
3. 2의 두개를 합쳐서 Estimator이라고 하는데 이것은 fit(), 과 predic()를 내부에서 구현하고 있다.
            
            cross_val_score()과 같은 evaluation 함수, GridsearchCV와 같은 하이퍼 파라미터 튜닝을 지원하는 클래스의 경우
            Estimator를 인자로 받는다.
            이 Estimator의 fit()과 predict()를 호출해서 평가를 하거나 하이퍼 파라미터를 수행한다.

사이킷런에서 비지도학습인 차원 축소, 클러스터링, 피처 추출 등을 구현한 클래스는 대부분 역시 대부분 fit()과 transform()을 적용한다.

비지도학습과 피처추출에서 fit()은 지도학습 fit()과 같이 학습을 의미하는 것이아니라 입력 데이터의 형태에 맞춰 데이터를 변환하기 위한 사전 구조를 맞추는 작업이다.

fit()으로 변환을 위한 사전 구조를 맞추면 이후 입력 데이터의 차원 변환, 클러스터링, 피처 추출 등의 실제 작업은 transfom()으로 수행한다.

사이킷런은 fit()과 transform()을 합친 fit_transform()도 제공한다.

일반적으로 머신러닝 구축하는 주요 프로세스는
            1. 피처의 가공, 변경, 추출을 수행하는 피처처리
            2. ML알고리즘 학습/예측 수행
            3. 모델 평가의 단계를 반복적으로 수행하는 것이다.

내장된 예제 데이터 세트: 사이킷런에는 데이터 세트가 내장되어 있다.
    저장된 데이터 세트는 일반적으로 딕셔너리 형태로 되어있다.
    키는 
        data: 피처의 데이터 세트를 가르킨다.
        target: 분류시 레이블 값, 회귀일 때는 숫자 결괏값 데이터 세트이다.
        target_names: 개별 레이블의 이름을 나타낸다.
        feature_names: 피처의 이름을 나타낸다.
        DESCR: 데이터 세트에 대한 설명과 각 피처의 설명을 나타낸다. 
        
data,target은 넘파이 타입이며, target_name, feature_name은 넘파이 도는 파이썬 리스트 타입이다. DESCR은 스트링 타입이다.

In [81]:
from sklearn.datasets import load_iris

iris_data=load_iris()
print(type(iris_data))

<class 'sklearn.utils.Bunch'>


Bunch 클래스는 파이썬의 딕셔너리 자료형과 유사하다. 
데이터 세트에 내장돼 있는 대부분의 데이터 세트는 대부분 딕셔너리 형태의 값을 반환한다.

In [82]:
keys=iris_data.keys()
print(keys)

dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names', 'filename'])
